src/lib/expression/correlated_parameter_expression.cpp

```diff
@@ -0,0 +1,77 @@
+#include "correlated_parameter_expression.hpp"
+
+#include <sstream>
+#include <string>
+#include <type_traits>
+
+#include "boost/functional/hash.hpp"
+
+#include "resolve_type.hpp"
+
+namespace opossum {
+
+CorrelatedParameterExpression::CorrelatedParameterExpression(const ParameterID parameter_id,
+                                                             const AbstractExpression& referenced_expression)
+    : AbstractParameterExpression(ParameterExpressionType::Correlated, parameter_id),
+      _referenced_expression_info(referenced_expression.data_type(), referenced_expression.is_nullable(),
+                                  referenced_expression.as_column_name()) {}
+
+CorrelatedParameterExpression::CorrelatedParameterExpression(const ParameterID parameter_id,
+                                                             const ReferencedExpressionInfo& referenced_expression_info)
+    : AbstractParameterExpression(ParameterExpressionType::Correlated, parameter_id),
+      _referenced_expression_info(referenced_expression_info) {}
+
+std::shared_ptr<AbstractExpression> CorrelatedParameterExpression::deep_copy() const {
+  return std::make_shared<CorrelatedParameterExpression>(parameter_id, _referenced_expression_info);
+}
+
+std::string CorrelatedParameterExpression::as_column_name() const {
+  std::stringstream stream;
+  stream << "Parameter[";
+  stream << "name=" << _referenced_expression_info.column_name << ";";
+  stream << "id=" << std::to_string(parameter_id);
+  stream << "]";
+
+  return stream.str();
+}
+
+DataType CorrelatedParameterExpression::data_type() const { return _referenced_expression_info.data_type; }
+
+bool CorrelatedParameterExpression::is_nullable() const { return _referenced_expression_info.nullable; }
+
+const std::optional<AllTypeVariant>& CorrelatedParameterExpression::value() const { return _value; }
+
+void CorrelatedParameterExpression::set_value(const std::optional<AllTypeVariant>& value) {
+  Assert(!value || data_type_from_all_type_variant(*value) == _referenced_expression_info.data_type ||
+             variant_is_null(*value),
+         "Invalid value assigned to CorrelatedParameterExpression");
+  _value = value;
+}
+
+bool CorrelatedParameterExpression::_shallow_equals(const AbstractExpression& expression) const {
+  const auto* parameter_expression_rhs = dynamic_cast<const CorrelatedParameterExpression*>(&expression);
+
+  return parameter_expression_rhs && parameter_id == parameter_expression_rhs->parameter_id &&
+         _referenced_expression_info == parameter_expression_rhs->_referenced_expression_info;
+}
+
+size_t CorrelatedParameterExpression::_on_hash() const {
+  auto hash = boost::hash_value(parameter_id.t);
```

Comment:
I don't like the `.t` part here. `static_cast<...>()` would be more verbose but the intent would be clearer.

Reply:
`static_cast<ParameterID::base_type>(parameter_id)`, u sure?

Reply:
It's ugly, but communicates its intent better than `.t`.

Reply:
kk